### ISSUE TITLE
Add ASUP analytics

### ISSUE THREAD

**Issue by cgrinds** [NetApp/harvest#81]
Similar to Harvest 1.6, Harvest2 should offer an option (ASUP, EMS, etc) to track # of installs, # pollers, etc.

Code that sends ASUP from Harvest 1.6
```
# Send AutoSupport with statistics about Harvest poller
sub send_autosupport_stats()
{
	
	# Get counts of nodes and volumes monitored
	# No point to continue if we don't have node instances
	my $nodes_count = keys %{$instance{"system:node"}};
	return 0 unless ($nodes_count > 0);
	my $vol_count = keys %{$instance{volume}};
	
	# Get OS
	my $distro = `cat /etc/os-release 2>/dev/null | grep PRETTY` || $Config{osname} || "unknown";
	# If we read os-release, what we need is the string after = optionally between single/double quotes
	$distro =~ s/^\s?PRETTY_NAME\s?=\s?['"]?(.*?)\s?['"]?.?$/$1/ms; # if $distro =~ /PRETTY/;

	# Runtime of worker
	my $runtime = `ps -ly --pid=$$ --no-headers 2>/dev/null` || "--";
	$runtime =~ s/.*\s(\d\d:\d\d:\d\d)\s.*/$1/ms;

	# Compose name for Harvest instance
	my $host_name = hostname();
	my $harvest_name = $connection{group} . "_" . $host_name;;

	# Stats about worker performance
	my $stats = "";
	for my $k ( qw (metrics fails skips plugin_time api_time last_time) )
	{
		$stats = $stats . ";" . $connection{statistics}{$k};
	}

	my $log_message = "HARVEST $VERSION [$harvest_name] [$distro] [$connection{product};$connection{host_version_generation};$connection{host_version_major}] [$nodes_count;$vol_count;$runtime] [$stats]";

	logger ("DEBUG", "[send_autosupport_stats] Composed AutuSupport log with statistics [$log_message]");

	my $server = $connection{server_obj};

	for my $node (keys %{$instance{"system:node"}})
	{
		my $node_name = $instance{"system:node"}{$node}{name};

		my $in = NaElement->new("autosupport-invoke");
		$in->child_add_string("force", "true");
		$in->child_add_string("message", $log_message);
		$in->child_add_string("node-name", $node_name);
		$in->child_add_string("type", "all");

		my $out = $server->invoke_elem($in);

		if ($out->results_status() eq "passed")
		{
			logger ("NORMAL", "[send_autosupport_stats] Sent AutoSupport Log Message for Node [$node_name] with statistics [$log_message].");
		}
		else
		{
			logger ("WARNING", "[send_autosupport_stats] Sending AutoSupport log message for Node [$node_name] failed with reason: $out->results_reason(). Message was [$log_message].");
		}
	}

	return 0;
}
```

**Comment by cgrinds** [comment]
Start simple and iterate:
* #nodes, #volumes, #svm 
* os
* kind of install
* harvest version

**Comment by vgratian** [comment]
## PTT0

I started implementing prototype - `PTT0` - in [](). Here are the details, any feedback, especially from Harvest users would be much appreciated!

## Motivation
It would help future development of Harvest if we have some idea about the size of deployment of Harvest and also some basic performance stats about Harvest outside of our development lab. This could be done my adding a feature to Harvest to send some telemetry data to NetApp using the ASUP messaging mechanism.

Harvest will only collect and emit this telemetry only with the explicit consent from its users. In PPT0, telemetry is only sent if the `send_asup_stats: true` (name might change) is added under the parameters of one or many pollers. By default, a poller will not collect/send telemetry.

Poller will collect data that give us some idea about it's performance, it will not collect any sensitive data such as the name of clusters, names of nodes, hostname of the local system. Below is a detailed overview of the telemetry collected in PTT0.

## Collected Data

There are 4 categories of data collected:

### HarvestInfo
General info Harvest version/build. This info is useful to understand the deployment of Harvest (e.g. rpm vs native installation)

| parameter            | source                           | description                                  | notes                  |
| ----------------------- | ------------------------------- | ----------------------------------------- | ---------------------- |
| `Version`             |  `cmd/harvest/version`   |                                                    |                   |
| `Release`            | same                              |                                                    |                   |
| `Commit`             | same                              |                                                    |                   |
| `BuildDate`          | same                             |                                                     |                   |
| `NumClusters`     | `harvest.yml`                 | number of pollers that poll an ONTAP system  |  partial guesswork (based on configured collectors)  |


### PlatformInfo
Information about the system where Harvest is Running

| parameter            | source                           | description                                  | notes                  |
| ----------------------- | ------------------------------- | ----------------------------------------- | ---------------------- |
| `OS`             |  `/etc/os-release`   |  Operating System             |                   |
| `CPUs`             |  `$ lscpu`   |  Number of CPUs             |                   |
| `Arch`             |  `$ lscpu`   |  CPU architecture             |                   |
| `MemoryKb`             |  `$ free --kilo`   |  Total Memory (RAM) Size             |                   |


### TargetInfo
Information about the target ONTAP system

| parameter            | source                           | description                                  | notes                  |
| ----------------------- | ------------------------------- | ----------------------------------------- | ---------------------- |
| `Version`             |  ZapiPerf collector   |              |                   |
| `Model`             |  same  |  `CDOT` or `7Mode`             |                   |
| `Ping`             |  poller metadata   |  average ping response time             |                   |


### InstanceInfo
Number of instances of 3 ZapiPerf objects: Node, SVM, Volume. This will give us some idea about the size of the Clusters monitored.

| parameter            | source                           | description                                  | notes                  |
| ----------------------- | ------------------------------- | ----------------------------------------- | ---------------------- |
| `Count`             |  ZapiPerf metadata   |  number of Instances            |                   |
| `DataPoints`             |  same  | average data points collected per poll            |                   |
| `PollTime`             |  same   |  average poll time of the collector             |                   |
| `ApiTime`             |  same   |  average API/wait time            |                   |
| `ParseTime`             |  same   |  average XML parse time            |                   |
| `PluginTime`             |  same   |  average Plugin time (if any)            |                   |

## Output

The collected data is rendered as JSON and is sent using the ASUP messaging. Open questions:
* Does the Harvest user need extra permissions?
* Use other methods to send? E.g. HTTP/email?

Example of the JSON data:

```json
{
  "Target": {
    "Version": "9.7.0",
    "Model": "CDOT",
    "Ping": 98.602
  },
  "Nodes": {
    "Count": 0,
    "DataPoints": 46,
    "PollTime": 500027,
    "ApiTime": 494896,
    "ParseTime": 3763,
    "PluginTime": 0
  },
  "SVMs": {
    "Count": 0,
    "DataPoints": 19,
    "PollTime": 450383,
    "ApiTime": 449627,
    "ParseTime": 425,
    "PluginTime": 0
  },
  "Volumes": {
    "Count": 0,
    "DataPoints": 230,
    "PollTime": 742318,
    "ApiTime": 500892,
    "ParseTime": 13092,
    "PluginTime": 104
  },
  "Platform": {
    "OS": "Manjaro Linux",
    "Arch": "x86_64",
    "MemoryKb": 7892262,
    "CPUs": 4
  },
  "Harvest": {
    "Version": "21.06.2416",
    "Release": "rc2",
    "Commit": "6a81edb",
    "BuildDate": "2021-06-24T16:35:04+0400",
    "NumClusters": 1
  }
}
````

**Comment by vgratian** [comment]
I just see `"Count": 0,` for the instances, will fix this.

**Comment by cgrinds** [comment]
Brian Ma had a good suggestion for next version. 
> Harvest should include how frequently it's querying data and which counters/config zapis

**Comment by cgrinds** [comment]
Including Prometheus and/or InfluxDB version would be helpful too for understand where to invest dashboard time

**Comment by Hardikl** [comment]
TODOs for ASUP feature:

Public side:
- [ ] payload fields structure (sample payload at below)
- [ ] schedule via poller to write in asup dir and exec asup binary 
- [ ] private binary to harvest repo
- [ ] License and legal updates/review
- [ ] SVM collection 
- [ ] Testing on all platform 

Private side:
- [ ] payload to asup flags 
- [ ] preprocessing with config and headers 
- [ ] connection to asup site 
- [ ] targz
- [ ] Logging in private binary to asup.log
- [ ] Limits – use something beside minio - engine.go


{
 "Target": {
  "Version": "9.8.0",
  "Model": "cdot",
  "Serial": "1-80-000011",
  "Ping": 0,
  "ClusterUuid": "cc476f88-9840-11eb-814a-xxxxxxxxxx"
 },
 "Nodes": {
  "Count": 1,
  "DataPoints": 16,
  "PollTime": 196714,
  "ApiTime": 195872,
  "ParseTime": 421,
  "PluginTime": 7,
  "NodeUuid": [
   "bf3b577e-983f-11eb-814a-xxxxxxxxxx"
  ]
 },
 "Volumes": {
  "Count": 18,
  "DataPoints": 3310,
  "PollTime": 910170,
  "ApiTime": 307509,
  "ParseTime": 18553,
  "PluginTime": 669
 },
 "Svms": {
  "Count": 0,
  "DataPoints": 0,
  "PollTime": 0,
  "ApiTime": 0,
  "ParseTime": 0,
  "PluginTime": 0
 },
 "Platform": {
  "OS": "darwin",
  "Arch": "darwin",
  "MemoryKb": 78864,
  "CPUs": 1
 },
 "Harvest": {
  "UUID": "a94fab28c249ac4c0c874a1f1cdec7b5xxxxxxxxxx",
  "Version": "21.08.1914",
  "Release": "v21.05.4",
  "Commit": "aac8950",
  "BuildDate": "2021-08-19T14:13:42+0530",
  "NumClusters": 1
 }
}

**Comment by cgrinds** [comment]
`21.08` includes everything but svm counters and minio limits

**Comment by cgrinds** [comment]
Included in 21.08. Closing
